### DISCUSSION TITLE
label smoothing as an optionally pre-processing step for model training

### DISCUSSION THREAD

**Discussion by ddbuscombe** [Doodleverse/segmentation_gym#52] [Ideas]
I propose that it would be possible to smooth labels on the fly during model training using a custom training loop or lambda layer in the model. The latter would be preferable

the smoothing would consist of a spatial-domain filter on one-hot encoded matrix slices, which are binary 0s and 1s

for example, a median filter 

what does this look like in keras?

`model.add(Lambda(lambda x: apply_median_filter(x, kernel_size)))`

?

**Comment by ebgoldstein** [comment]
Seems like we could use: https://www.tensorflow.org/addons/api_docs/python/tfa/image/median_filter2d.

this is going to happen to the labels or the model generated predictions? if labels, should we put in `make_dataset` and not the model?

if it was going to be the model as a Lambda layer, we can do something like:
```
#define a median filer
def median_filter(x):
         x_filtered = tfa.image.median_filter2d(x)
        #do anything else we want to do
        return x_filtered
```
and then:
`model.add(tf.keras.layers.Lambda(median_filter))`

**Reply by ddbuscombe** [reply]
Was thinking this would only apply to the labels. It could go either in make_datasets or train_model. I vote the latter, because then it happens on the fly, and the stored labels are unaffected. It would facilitate easier comparison between models trained on smoothed versus original, non-smoothed, labels, because it would require only one dataset
